### PR TITLE
Cleanup services/networks only on successful test run

### PR DESCRIPTION
**Pull request by sanimej**
cc @dhiltgen. This should help to identify the reason for intermittent failure in some tests.

Signed-off-by: Santhosh Manohar <santhosh@docker.com>

### CODE REVIEW

**Comment by dperny** [comment]
I wrote this if you want to use it. I'm reasonably sure it works.

PR LGTM whether you use it or not. 

```go
// ExecIfPassing executes the given func only if the test has passed
//
// f is generally a closure of some sort. go has no good delayed evaluation
// primitives but passing a bare function is generally a good way to do it.
//
// It is used so that some defered functions, especially those used to clean up
// test objects, are only run if a test passes. This allows test writers to
// skip cleanup so that the objects associated with failing tests can be
// investigated later for trouble
func ExecIfPassing(t *testing.T, f func()) {
	if t.Failed() {
		return
	}
	f()
}
```

File: tests/network_test.go [hunk 1]
```diff
@@ -37,7 +37,12 @@ func TestServiceDiscovery(t *testing.T) {
 	}
 	_, err = cli.NetworkCreate(testContext, nwName, nc)
 	require.NoError(t, err, "Error creating overlay network %s", nwName)
-	defer cli.NetworkRemove(testContext, nwName)
+	defer func() {
+		if t.Failed() {
+			return
+		}
+		cli.NetworkRemove(testContext, nwName)
```
Comment (dperny) [inline tests/network_test.go]: I didn't realize this earlier, but there is a possible case where the test passes right before the testContext times out, which would result in this network remove never getting executed (because the context will already have been canceled).

I should have noted this in an earlier review. I probably made the same mistake in my code.

**Comment by sanimej** [comment]
I like the approach of having hiding the `t.Failed()` check inside a utility function. As a convention if we call this function in all the places it avoids the accidental forgetting of not checking for it and also the boiler-plate checks everywhere.. updating the PR with this change.

File: tests/network_test.go (hunk 1 above)
Comment (sanimej) [inline tests/network_test.go]: Either the context timeout will lead to test failure or the test will pass. It can't be both, correct ? If the context times out thats ok. May be we can increase the timeout for that case.

Comment (dperny) [inline tests/network_test.go]: No, actually. If that context times out, it'll cause subsequent context-dependent calls to be skipped, which will cause failures and then cause the test to end. But the context timing out by itself doesn't cause this. So there exists a possible race in which the context times out after the test is completed but before these API calls are executed, which causes them to be skipped. It's not a major issue, nor is it a problem with your PR, it's just something I realized while reviewing.

Comment (sanimej) [inline tests/network_test.go]: Ok, got it.. the context timeout should be per API that uses the context. I haven't paid close attention to the context usage in e2e. We can do it and make necessary changes in a separate PR.

Comment (dperny) [inline tests/network_test.go]: Yeah, this is something we need to do in the future, but it's not a blocker for this PR.

**Comment by dperny** [comment]
Reiterating, LGTM.

**Comment by sanimej** [comment]
Got two LGTMs.. merging.
